Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,446 @@
+import Batteries.Data.Array
+import Batteries.Data.List.Basic
+import Batteries.Data.List.Lemmas
+
+namespace Vector
+
+open Array Lean
+
+/--
+  `Vector α n` is an `Array α` whose size is statically fixed to `n`
+-/
+structure Vector (α : Type u) (n : Nat) where
+  /--Internally, a vector is stored as an array for fast access-/
+  toArray : Array α
+  /--`size_eq` fixes the size of `toArray` statically-/
+  size_eq : toArray.size = n
+deriving Repr, BEq, DecidableEq
+
+
+/--Syntax for `Vector α n`-/
+syntax "#v[" withoutPosition(sepBy(term, ", ")) "]" : term
+
+macro_rules
+  | `(#v[ $elems,* ]) => `(Vector.mk (n := $(quote elems.getElems.size)) #[$elems,*] rfl)
+
+#check #v[1,2,3,3,4,4]
+
+/--
+  `Vector.size` gives the size of a vector.
+-/
+def size (_ : Vector α n) : Nat := n
+
+
+/--
+  `Vector.nil` produces an empty vector
+-/
+@[match_pattern]
+def nil : Vector α 0 := ⟨Array.empty, by rfl⟩
+
+/--
+  `Vector.cons` is the list cons pattern for vectors
+-/
+@[match_pattern]
+def cons (x : α) (xs : Vector α n) : Vector α (Nat.succ n) :=
+  match xs with
+  | ⟨{data := xs'}, h⟩ => ⟨{data := x :: xs'}, congrArg Nat.succ h⟩
+
+/--
+  `Vector.head` produces the head of a vector
+-/
+def head (v : Vector α (Nat.succ n)) :=
+  match v with
+  | ⟨{data := a :: _}, _⟩ => a
+
+/--
+  `Vector.tail` produces the tail of a vector
+-/
+def tail (v : Vector α n) : Vector α (n-1):=
```

Comment:
Should use `Array.feraseIdx`.

Reply:
Thanks I added it.

Reply:
Suggestion
```lean
/--
  Remove the element at a given index from a vector without bounds checks, using a `Fin` index.

  This function takes worst case O(n) time because it has to backshift all elements at positions 
  greater than i.
-/
def feraseIdx : Vector α n → Fin n → Vector α (n-1)
  | ⟨a, h⟩, ⟨i, hi⟩ => ⟨a.feraseIdx ⟨i, h ▸ hi⟩, h ▸ a.size_feraseIdx ..⟩

/--
  `Vector.tail` produces the tail of a vector
-/
def tail (v : Vector α n) : Vector α (n-1) := 
  match n with
  | 0 => v
  | _+1 => v.feraseIdx 0
```

Reply:
I just noticed! To follow the Array API, `eraseIdx` would use a `Nat` index instead of a `Fin` index.

Reply:
The array version returns an array of the same size if `i < n` is not true. This doesn't make sense for `Vector a (n-1)`

Reply:
Done. But I had to add a type class constraint `[Inhabited \alpha]` because panic somehow requires this and unlike `swapAt!` there is no obvious candidate to show `Inhabited \alpha` which leads to `Inhabited (Vector \alpha (n-1))`

Reply:
How about adding `let _ : Inhabited (Vector α (n-1)) := ⟨v.tail⟩`?

Reply:
This works:
```lean
def eraseIdx! [Inhabited α] (v : Vector α n) (i : Nat) : Vector α (n-1) :=
    if h : i < n then
      feraseIdx v ⟨i,h⟩
    else
      let _ : Inhabited (Vector α (n-1)) := ⟨v.tail⟩
      panic! s!"Index {i} is out of bounds"
```